scripts/Muon/GUI/Common/muon_context.py

```diff
@@ -4,17 +4,136 @@
+                    self.group_pair_context[pair_name].update_asymmetry_workspace(pair_asymmetry_workspace, run, rebin=True)
+
+    def calculate_all_groups(self):
+        for run in self._data_context.current_runs:
+            for group_name in self._group_pair_context.group_names:
+                group_workspace, group_asymmetry = self.calculate_group(group_name, run)
+                self.group_pair_context[group_name].update_workspaces(run, group_workspace, group_asymmetry, rebin=False)
+
+                if self._do_rebin():
+                    group_workspace, group_asymmetry = self.calculate_group(group_name, run, rebin=True)
+                    self.group_pair_context[group_name].update_workspaces(run, group_workspace, group_asymmetry, rebin=True)
+
+    def update_current_data(self):
+        # Update the current data; resetting the groups and pairs to their default values
+        if len(self.data_context.current_runs) > 0:
+            self.data_context.update_current_data()
+
+            if not self.group_pair_context.groups:
+                self.group_pair_context.reset_group_and_pairs_to_default(self.data_context.current_workspace,
+                                                                         self.data_context.instrument,
+                                                                         self.data_context.main_field_direction)
+        else:
+            self.data_context.clear()
+
+    def show_raw_data(self):
+        for run in self.data_context.current_runs:
+            run_string = run_list_to_string(run)
+            loaded_workspace = self.data_context._loaded_data.get_data(run=run, instrument=self.data_context.instrument)['workspace'][
+                'OutputWorkspace']
+            directory = get_base_data_directory(self, run_string) + get_raw_data_directory(self, run_string)
+
+            if len(loaded_workspace) > 1:
+                # Multi-period data
+                for i, single_ws in enumerate(loaded_workspace):
+                    name = directory + get_raw_data_workspace_name(self, run_string, period=str(i + 1))
+                    single_ws.show(name)
+            else:
+                # Single period data
+                name = directory + get_raw_data_workspace_name(self, run_string)
+                loaded_workspace[0].show(name)
+
+    def _do_rebin(self):
+        return (self.gui_context['RebinType'] == 'Fixed' and
+                'RebinFixed' in self.gui_context and self.gui_context['RebinFixed']) or\
+               (self.gui_context['RebinType'] == 'Variable' and
+                'RebinVariable' in self.gui_context and self.gui_context['RebinVariable'])
+
+    @property
+    def first_good_data(self):
+        if self.gui_context['FirstGoodDataFromFile']:
+            return self.data_context.get_loaded_data_for_run(self.current_runs[-1])["FirstGoodData"]
+        else:
+            return self.gui_context['FirstGoodData']
```

Comment:
does this file have tests?

Reply:
Yes they are here https://github.com/mantidproject/mantid/blob/a0dd9776ceac4f2ca557ed89193a60c0ee3cc64c/scripts/test/Muon/muon_context_test.py